File: Solutions/Cpp/product_of_array_except_self.cpp
/*
    [MEDIUM]
    238. Product of Array Except Self

    Concepts:
    - array
    - prefix sum

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 42.26 MB  [Beats 16.73%]
*/

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        std::vector<int> pref(n, 1);
        std::vector<int> suff(n, 1);
        std::vector<int> res(n, 1);

        for (int i = 1; i < n; i++)
            pref[i] = pref[i-1] * nums[i-1];

        for (int i = n-2; i > -1; i--)
            suff[i] = suff[i+1] * nums[i+1];

        for (int i = 0; i < n; i++)
            res[i] = pref[i] * suff[i];

        return res;
    }
};

File: Solutions/Rust/product_of_array_except_self.rs
/*
    [MEDIUM]
    238. Product of Array Except Self

    Concepts:
    - array
    - prefix sum

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 3.88 MB   [Beats 9.74%]
*/
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n: usize = nums.len();
        let mut pref: Vec<i32> = Vec::with_capacity(n);
        for _ in 0..n { pref.push(1); }
        let mut suff: Vec<i32> = pref.clone();
        let mut res: Vec<i32> = pref.clone();

        for i in 1..n {
            pref[i] = pref[i-1] * nums[i-1];
        }
        for i in (0..n-1).rev() {
            suff[i] = suff[i+1] * nums[i+1];
        }
        for i in 0..n {
            res[i] = pref[i] * suff[i];
        }
        
        return res;
    }
}